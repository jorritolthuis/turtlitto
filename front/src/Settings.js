import React from "react";
import PropTypes from "prop-types";
import Turtle from "./Turtle";
import TurtleEnableBar from "./TurtleEnableBar";

/**
 * Show the settings of all turtles.
 * Author: H.E. van der Laan
 *
 * props:
 * - turtles: an array of Turtles
 * - onTurtleEnableChange: function to call when the turtle enable button is pressed
 */
const Settings = props => {
  const { turtles, onTurtleEnableChange } = props;
  return (
    <div>
      <TurtleEnableBar
        turtles={turtles.map(turtle => {
          return {
            id: turtle.id,
            enabled: turtle.enabled
          };
        })}
        onTurtleEnableChange={onTurtleEnableChange}
      />
      {turtles
        .filter(turtle => turtle.enabled)
        .map(turtle => (
          <Turtle
            key={turtle.id}
            turtle={turtle}
            editable
            onChange={(changedProp, newValue) => {} /* TODO: turtle update */}
          />
        ))}
    </div>
  );
};

Settings.propTypes = {
  onTurtleEnableChange: PropTypes.func.isRequired,
  turtles: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      enabled: PropTypes.bool.isRequired,
      battery: PropTypes.number.isRequired,
      home: PropTypes.string.isRequired,
      role: PropTypes.string.isRequired,
      team: PropTypes.string.isRequired
    })
  ).isRequired
};

export default Settings;
