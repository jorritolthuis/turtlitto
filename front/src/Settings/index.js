import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Dropdown from "../Dropdown";
import TurtleList from "../TurtleList";
import sendToServer from "../sendToServer";

/**
 * Show the settings of all turtles.
 * Author: H.E. van der Laan
 * Author: T.T.P. Franken
 *
 * props:
 * - turtles: an array of Turtles
 * - session: a string which holds the password needed to connect to the SRRS
 */

const CONFIG_VALUES = [
  "Role assigner on",
  "Role assigner off",
  "Pass demo",
  "Penalty mode",
  "Ball Handling demo"
];

const COMMAND_VALUES = {
  "Role assigner on": "role_assigner_on",
  "Role assigner off": "role_assigner_off",
  "Pass demo": "pass_demo",
  "Penalty mode": "penalty_demo",
  "Ball Handling demo": "ball_handling_demo"
};

const Settings = props => {
  const { turtles, session } = props;
  return (
    <SettingsWrapper>
      <RoleDropdown
        id={"settings_role-dropdown"}
        currentValue={"Whatever"}
        values={CONFIG_VALUES}
        onChange={value => {
          sendToServer(COMMAND_VALUES[value], "command", session);
        }}
        enabled={true}
      />
      <TurtleList turtles={turtles} session={session} />
    </SettingsWrapper>
  );
};

Settings.propTypes = {
  turtles: PropTypes.objectOf(
    PropTypes.shape({
      enabled: PropTypes.bool.isRequired,
      batteryvoltage: PropTypes.number.isRequired,
      homegoal: PropTypes.string.isRequired,
      role: PropTypes.string.isRequired,
      teamcolor: PropTypes.string.isRequired
    })
  ).isRequired
};

const RoleDropdown = styled(Dropdown)`
  margin-left: auto;
  margin-right: auto;
  height: 4rem;
`;

const SettingsWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export default Settings;
