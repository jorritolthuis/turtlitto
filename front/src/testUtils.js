import React from "react";
import { ThemeProvider } from "styled-components";
import theme from "./theme";
import { shallow, mount } from "enzyme";

export const shallowWithTheme = tree => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallow(tree, { context });
};

export const mountWithTheme = tree => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();

  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes
  });
};
