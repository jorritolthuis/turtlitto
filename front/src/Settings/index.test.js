import React from "react";
import { shallow } from "enzyme";
import Settings from ".";
import { mountWithTheme } from "../testUtils";

it("renders without crashing", () => {
  const wrapper = mountWithTheme(
    <Settings
      onChange={() => {}}
      command="pass_demo"
      turtles={{
        1: {
          enabled: true,
          batteryvoltage: 66,
          homegoal: "yellow",
          role: "inactive",
          teamcolor: "magenta"
        },
        2: {
          enabled: false,
          batteryvoltage: 42,
          homegoal: "yellow",
          role: "inactive",
          teamcolor: "magenta"
        }
      }}
      onTurtleEnableChange={() => {}}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

describe("When the role assigner dropdown is changed to", () => {
  describe("the user changes the the role assigner dropdown to", () => {
    const realFetch = global.fetch;
    let wrapper = null;

    beforeEach(() => {
      wrapper = shallow(
        <Settings
          turtles={{
            1: {
              enabled: true,
              batteryvoltage: 66,
              homegoal: "yellow",
              role: "inactive",
              teamcolor: "magenta"
            },
            2: {
              enabled: false,
              batteryvoltage: 42,
              homegoal: "yellow",
              role: "inactive",
              teamcolor: "magenta"
            }
          }}
        />
      );
      const l = window.location;
      global.fetch = jest.fn().mockImplementation((url, params) => {
        expect(url).toBe(`${l.protocol}//${l.host}/api/v1/command`);
        expect(params).toMatchSnapshot();
        return Promise.resolve({ ok: true });
      });
    });

    afterEach(() => {
      global.fetch = realFetch;
    });

    it("'Role Assigner on', it should pass 'role_assigner_on' to sendToServer", () => {
      wrapper
        .find("#settings_role-dropdown")
        .simulate("change", "Role assigner on");
    });

    it("'Role Assigner off', it should pass 'role_assigner_off' to sendToServer", () => {
      wrapper
        .find("#settings_role-dropdown")
        .simulate("change", "Role assigner off");
    });

    it("'Pass demo', it should pass 'pass_demo' to sendToServer", () => {
      wrapper.find("#settings_role-dropdown").simulate("change", "Pass demo");
    });

    it("'Penalty mode', it should pass 'penalty_demo' to sendToServer", () => {
      wrapper
        .find("#settings_role-dropdown")
        .simulate("change", "Penalty mode");
    });

    it("'Ball Handling demo', it should pass 'ball_handling_demo' to sendToServer", () => {
      wrapper
        .find("#settings_role-dropdown")
        .simulate("change", "Ball Handling demo");
    });
  });
});
