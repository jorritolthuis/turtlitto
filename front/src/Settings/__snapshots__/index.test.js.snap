// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Ball Handling demo', it should pass 'ball_handling_demo' to sendToServer 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Ball Handling demo', it should pass 'ball_handling_demo' to sendToServer 2`] = `
Object {
  "body": "\\"ball_handling_demo\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Pass demo', it should pass 'pass_demo' to sendToServer 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Pass demo', it should pass 'pass_demo' to sendToServer 2`] = `
Object {
  "body": "\\"pass_demo\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Penalty mode', it should pass 'penalty_demo' to sendToServer 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Penalty mode', it should pass 'penalty_demo' to sendToServer 2`] = `
Object {
  "body": "\\"penalty_demo\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Role Assigner off', it should pass 'role_assigner_off' to sendToServer 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Role Assigner off', it should pass 'role_assigner_off' to sendToServer 2`] = `
Object {
  "body": "\\"role_assigner_off\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Role Assigner on', it should pass 'role_assigner_on' to sendToServer 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;

exports[`When the role assigner dropdown is changed to the user changes the the role assigner dropdown to 'Role Assigner on', it should pass 'role_assigner_on' to sendToServer 2`] = `
Object {
  "body": "\\"role_assigner_on\\"",
  "headers": Object {},
  "method": "PUT",
}
`;

exports[`renders without crashing 1`] = `
<styled.div>
  <Styled(Dropdown)
    currentValue="Whatever"
    enabled={true}
    id="settings_role-dropdown"
    onChange={[Function]}
    values={
      Array [
        "Role assigner on",
        "Role assigner off",
        "Pass demo",
        "Penalty mode",
        "Ball Handling demo",
      ]
    }
  />
  <TurtleList
    turtles={
      Object {
        "1": Object {
          "batteryvoltage": 66,
          "enabled": true,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
        "2": Object {
          "batteryvoltage": 42,
          "enabled": false,
          "homegoal": "Yellow home",
          "role": "INACTIVE",
          "teamcolor": "Magenta",
        },
      }
    }
  />
</styled.div>
`;
